### PR TITLE
develop > release branch v1.0.0

### PR DESCRIPTION
**Pull request by melferink**
Release of v1.0.0 for NextflowONT for SMA workflow

### CODE REVIEW

File: Modules/GATK/4.2.1.0/FilterVCFs.nf [hunk 1]
```diff
@@ -0,0 +1,22 @@
+process FilterVcfs {
+    tag {"GATK FilterVCFs ${sample_id}"}
+    label 'GATK_4_2_1_0'
+    label 'GATK_4_2_1_0_MergeVcfs'
```
Comment (rernst) [inline Modules/GATK/4.2.1.0/FilterVCFs.nf]: Merge or Filter?

File: Modules/Longshot/0.4.1/Phase.nf [hunk 2]
```diff
@@ -0,0 +1,23 @@
+process LongshotPhase {
```
Comment (rernst) [inline Modules/Longshot/0.4.1/Phase.nf]: Why did you add 'Phase'?

File: Modules/Minimap2/2.26--he4a0461_1/Mapping.nf [hunk 3]
```diff
@@ -0,0 +1,19 @@
+process Mapping {
+    tag {"Minimap2 ${fastq.simpleName}"}
+    label 'Minimap_2_26'
```
Comment (rernst) [inline Modules/Minimap2/2.26--he4a0461_1/Mapping.nf]: Add Minimap mapping label?

Comment (rernst) [reply Modules/Minimap2/2.26--he4a0461_1/Mapping.nf]: Or remove 'mapping'

File: Modules/Sambamba/1.0.0/Filter.nf [hunk 4]
```diff
@@ -0,0 +1,65 @@
+process FilterROI {
```
Comment (rernst) [inline Modules/Sambamba/1.0.0/Filter.nf]: If you follow our nextflow guidelines this file / process / tag / label/ etc should be named 'view', since you are using the view command.

File: Modules/Sambamba/1.0.0/Filter.nf [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    script:
+        roi_slice = ""
+        if( params.roi )
+            roi_slice = " $params.roi "
+        """
+        sambamba view -t ${task.cpus} -f bam \
+        ${bam_file} \
+        $roi_slice \
+        -o  ${bam_file.simpleName}_roi.bam
```
Comment (rernst) [inline Modules/Sambamba/1.0.0/Filter.nf]: This works probably fine but I would try to put 'named' arguments/options first before positional arguments.

File: Modules/Sambamba/1.0.0/Filter.nf (hunk 4 above)
Comment (rernst) [reply Modules/Sambamba/1.0.0/Filter.nf]: Like viewsort

File: Modules/Sniffles2/2.2--pyhdfd78af_0/VariantCaller.nf [hunk 6]
```diff
@@ -0,0 +1,22 @@
+process VariantCaller {
```
Comment (rernst) [inline Modules/Sniffles2/2.2--pyhdfd78af_0/VariantCaller.nf]: Just 'sniffles'?

File: Modules/Tabix/1.11/Index.nf [hunk 7]
```diff
@@ -0,0 +1,19 @@
+process ZipIndex {
```
Comment (rernst) [inline Modules/Tabix/1.11/Index.nf]: In this case I would rename the file to ZipIndex as well.

File: Modules/Utils/GuppyBasecalling.nf [hunk 8]
```diff
@@ -0,0 +1,29 @@
+process ReBasecallingGuppy{
```
Comment (rernst) [inline Modules/Utils/GuppyBasecalling.nf:1]: Why not include in a 'Guppy' tool folder?

File: Modules/Whatshap/1.7/Haplotag.nf [hunk 9]
```diff
@@ -0,0 +1,23 @@
+process Haplotag {
+    tag {"Whatshap_Phase Minimap2 ${bam_file}"}
```
Comment (rernst) [inline Modules/Whatshap/1.7/Haplotag.nf]: Tag from Phase.nf?

File: Modules/duplex_tools/0.2.17/FilterPairs.nf [hunk 10]
```diff
@@ -0,0 +1,20 @@
+process FilterPairs {
```
Comment (rernst) [inline Modules/duplex_tools/0.2.17/FilterPairs.nf]: We use CamelCase for tools (folder name).

File: README.md [hunk 11]
```diff
@@ -1,2 +1,61 @@
-# NextflowONT
-Nextflow Oxford Nanopore Technologies workflow
+# NextflowSMA
+Workflow SMA for Oxford Nanopore Technologies sequencing data\
+Note: this workflow hase been tested on R9.4.1 sequencing data with guppy_6.1.2 basecalling only.\
+Note: current workflow setup has been tested on Rocky Linux 8 combined with Slurm Workload Manager.
```
Comment (rernst) [inline README.md]: Add a `requirements` section?

Comment (rernst) [reply README.md]: And as a list -> remove the `\`.

File: README.md [hunk 12]
```diff
@@ -1,2 +1,61 @@
...
+
+#### Install OpenJDK and Nextflow
```
Comment (rernst) [inline README.md]: Maybe add some explanation what this install scripts does?

File: README.md [hunk 13]
```diff
@@ -1,2 +1,61 @@
...
+```bash
+sh install.sh
+```
+
+#### Configure paths before use
+SMA.config:
+<pre>
+  genome_fasta          	full path to reference genome fasta (.fasta/.fa/.fna). note that  the reference genome needs an index (.fai) and a dictionary (.dict)
```
Comment (rernst) [inline README.md:28]: How do you generate these files?

File: README.md [hunk 14]
```diff
@@ -1,2 +1,61 @@
...
+Note: current workflow setup has been tested on Rocky Linux 8 combined with Slurm Workload Manager.
+
+#### Install OpenJDK and Nextflow
+```bash
+sh install.sh
+```
+
+#### Configure paths before use
+SMA.config:
+<pre>
```
Comment (rernst) [inline README.md:27]: Using this `pre` block we need to scroll to read all the instructions, not sure what a better way of formatting is though. Maybe just a list:
- `genome_fasta`: 'description'
  - 'how to generate'

File: README.md [hunk 15]
```diff
@@ -1,2 +1,61 @@
...
+  * rebase              assumes raw data folder with fast5_pass/ and fast5_fail/ subfolders.
+  * bam(_remap)         assumes Guppy output folder with subfolder pass/ in which BAM files are located.
+  * bam_single(_remap)  assumes a single BAM file per input folder.
+</pre>
+##### --method
+<pre>
+  * rebase              start from FAST5 raw data ouput folder and include re-basecalling.
+  * bam                 start from Guppy folder including bam files.
+  * bam_remap           start from Guppy folder including bam files, but perform remapping to genome in config.
+  * bam_single          start from single bam file (should be only bam in the folder) without any Guppy/ONT information.
```
Comment (rernst) [inline README.md]: Rewrite: should be only bam in the folder -> Make sure only one bam file is present in the input path folder? Or something else a little bit more descriptive.

File: README.md (hunk 12 above)
Comment (rernst) [inline README.md]: Why 4 `#` not just 2? Leave the layout to the markdown parser?

File: README.md [hunk 16]
```diff
@@ -1,2 +1,61 @@
...
+  guppy_basecaller_path 	full path to guppy_basecaller executable
+  guppy_basecaller_config 	full path to basecalling model file to be used (.cfg)
+</pre>
+
+#### Running SMA workflow
+```bash
+nextflow run SMA.nf -c SMA.config --input_path <input_path> --outdir <output_dir_path> --start <bam|rebase> --method <method> --email <email> [-profile slurm]
+```
+
+##### --input_path
```
Comment (rernst) [inline README.md]: The way you write the `--input_path` and `--method` down in a similar way is a bit confusing. Input_path is a path not rebase / bam, etc. While the methods are actually the options with a *. 

So I would combine the two:

## Method and input_path
Depending on the chosen method, input_path assumes different file input formats.

### Rebase
Start from FAST5 raw data ouput folder and include re-basecalling.
- method = rebase
- input_path = raw data folder with fast5_pass/ and fast5_fail/ subfolders.

etc.

File: README.md [hunk 17]
```diff
@@ -1,2 +1,61 @@
...
+<pre>
+Input path is full path to either FAST5 or BAM file, depending on method (see below).
+  * rebase              assumes raw data folder with fast5_pass/ and fast5_fail/ subfolders.
+  * bam(_remap)         assumes Guppy output folder with subfolder pass/ in which BAM files are located.
+  * bam_single(_remap)  assumes a single BAM file per input folder.
+</pre>
+##### --method
```
Comment (rernst) [inline README.md]: I think this must be 'start'?

File: README.md [hunk 18]
```diff
@@ -1,2 +1,61 @@
...
+  * bam_single          start from single bam file (should be only bam in the folder) without any Guppy/ONT information.
+  * bam_single_remap    start from single bam file (should be only bam in the folder) without any Guppy/ONT information and perform remapping.
+</pre>
+
+| Method | Description | Optional parameters needed|
+| --- | :--- | :--- |
+|wgs|perform whole genome mapping + longshot phasing|none|
+|SMA_adaptive|perform whole genome mapping (if required) +  SMA Variant calling + haplotype phasing|--ploidy \<SMN2 copy number\>|
+
+in which:
```
Comment (rernst) [inline README.md]: I do not understand the 'in which:' part.

File: SMA.config [hunk 19]
```diff
@@ -0,0 +1,448 @@
...
+    //VCF annotation of SNV and INDELs called with GATK or Clair3
+    gatk_snp_filter = "--filter-name SNP_LowQualityDepth --filter-expression 'QD < 2.0' --filter-name SNP_MappingQuality --filter-expression 'MQ < 40.0' --filter-name SNP_StrandBias --filter-expression 'FS > 10.0' --filter-name Ref_Homopolymer --filter-expression 'RegionRef > 0'"
+    gatk_snp_cluster = "--cluster-size 3 --cluster-window-size 35"
+    gatk_indel_filter = "--filter-name INDEL_LowQualityDepth --filter-expression 'QD < 2.0' --filter-name INDEL_StrandBias --filter-expression 'FS > 200.0' --filter-name --filter-name --filter-expression 'ReadPosRankSum < -20.0'"
+    clair3_snp_cluster = "--cluster-size 3 --cluster-window-size 35"
+    clair3_snp_filter = "--filter-name Ref_Homopolymer --filter-expression 'RegionRef > 0'"
+    clair3_indel_filter = "--filter-name Ref_Homopolymer --filter-expression 'RegionRef > 0'"
+    // Region of interest for BAM > fastq is default empty. Parameter here enables possibility to select specific region (i.e. SMA locus). roi = chr:start-stop (e.g chr5:71274893-71447410)
+    roi_fastq = ""
```
Comment (rernst) [inline SMA.config]: Add some lines to seperate things? Like you did in the first rows.

File: SMA.config [hunk 20]
```diff
@@ -0,0 +1,448 @@
...
+
+    //General cluster options (Slurm)
+    cluster_options = "--mail-user $params.email --mail-type FAIL"
```
Comment (rernst) [inline SMA.config]: This is really system dependent, I am not sure whether you should include this.

File: SMA.config [hunk 21]
```diff
@@ -0,0 +1,448 @@
...
+}
+
+process {
+
+    withLabel: ReBasecallingGuppy {
+        cpus = 2
+        memory = { 20.GB * task.attempt }
+        time = { (3.m * fast5) * task.attempt }
+        queue = 'gpu'
+        clusterOptions = "$params.cluster_options --gres=tmpspace:32G  --gres=gpu:2g.20gb:1 "
```
Comment (rernst) [inline SMA.config]: See comment on line 39

File: SMA.config [hunk 22]
```diff
@@ -0,0 +1,448 @@
...
+            autoMounts = true
+            cacheDir = '{path}'
+        }
+
+        executor {
+            queueSize = 1000
+            pollInterval = '1min'
+            queueStatInterval = '5min'
+            submitRatelimit = '10sec'
+        }
```
Comment (rernst) [inline SMA.config]: This is UMCU HPC specific

File: SMA.config [hunk 23]
```diff
@@ -0,0 +1,448 @@
...
+            queue = 'cpu'
+            clusterOptions = "$params.cluster_options"
+
+            errorStrategy = 'retry'
+            maxRetries = 1
+        }
+
+        singularity {
+            enabled = true
+            runOptions = '-B /hpc:/hpc -B $TMPDIR:$TMPDIR'
```
Comment (rernst) [inline SMA.config]: hpc path

File: SMA.config [hunk 24]
```diff
@@ -0,0 +1,448 @@
...
+    fields = 'task_id,hash,native_id,process,tag,name,status,exit,module,container,cpus,time,disk,memory,attempt,submit,start,complete,duration,realtime,queue,%cpu,%mem,rss,vmem,peak_rss,peak_vmem,rchar,wchar,syscr,syscw,read_bytes,write_bytes,vol_ctxt,inv_ctxt'
+}
+
+timeline {
+    enabled = true
+    overwrite = true
+    file = "$params.outdir/log/nextflow_timeline.html"
+}
+
+profiles {
```
Comment (rernst) [inline SMA.config]: I am not sure if you should include any profiles...

File: SMA.nf [hunk 25]
```diff
@@ -0,0 +1,375 @@
+#!/usr/bin/env nextflow
+nextflow.enable.dsl=2
```
Comment (rernst) [inline SMA.nf]: It works on the latest nextflow right? This is not needed anymore.

`New in version 22.03.0-edge: DSL2 is the default DSL version.`

File: SMA.nf [hunk 26]
```diff
@@ -0,0 +1,375 @@
...
+include { ZipIndex as Tabix_Zip_Index_Bed } from './Modules/Tabix/1.11/Index.nf'
+include { ZipIndex as Tabix_Zip_Index_Region } from './Modules/Tabix/1.11/Index.nf'
+include { ZipIndex as Tabix_Zip_Index_Bcftools_Clair3_Bed } from './Modules/Tabix/1.11/Index.nf'
+include { ZipIndex as Tabix_Zip_Index_Bcftools_Clair3_Region } from './Modules/Tabix/1.11/Index.nf'
+include { ZipIndex as Tabix_Zip_Index_Bcftools_Bed } from './Modules/Tabix/1.11/Index.nf'
+include { ZipIndex as Tabix_Zip_Index_Bcftools_Region } from './Modules/Tabix/1.11/Index.nf'
+
+
+def analysis_id = params.outdir.split('/')[-1]
+sample_id = params.sample_id
```
Comment (rernst) [inline SMA.nf]: Just use `params.sample_id` in the code? Like you also do for `input_path`

File: SMA.nf [hunk 27]
```diff
@@ -0,0 +1,375 @@
...
+        summary_file = ReBasecallingGuppy.out.map{fastq_files, fast5_files, all_files, bam_files, summary_file -> summary_file}
+    }
+    else if( params.start == 'bam_remap' ){
+        // Get fast5 and mapped bams from input folder
+        bam_files = Channel.fromPath(params.input_path +  "/pass/*.bam").toList()
+        summary_file = Channel.fromPath(params.input_path +  "/sequencing_summary.txt").toList()
+    }
+    else{
+        error  """
+            Invalid alignment mode: ${params.start}.
```
Comment (rernst) [inline SMA.nf]: `Invalid start mode: <mode>. This should be:`

File: SMA.nf [hunk 28]
```diff
@@ -0,0 +1,375 @@
...
+
+
+    if( params.start == 'bam_single' ||  params.start == 'bam_single_remap' ){
+        // Index MergeSort BAM
+        Sambamba_Index_Merge(bam_file.map{bam_file -> [sample_id, bam_file]})
+
+        // Filter for minimum readlength
+        Sambamba_Filter_Condition(bam_file.combine(Sambamba_Index_Merge.out.map{sample_id, bai_file -> bai_file}))
+
+        bam_file = Sambamba_Filter_Condition.out
```
Comment (rernst) [inline SMA.nf:123]: Be careful with reusing variable names. Nextflow is not linear...

File: SMA.nf [hunk 29]
```diff
@@ -0,0 +1,375 @@
...
+        // Index MergeSort BAM
+        Sambamba_Index_Merge(Samtools_Merge.out)
+
+        // Filter for minimum readlength
+        Sambamba_Filter_Condition(Samtools_Merge.out
+            .map{ sample_id, bam_file -> bam_file }
+            .combine(Sambamba_Index_Merge.out
+                .map{ sample_id, bai_file -> bai_file }
+            )
+        )
```
Comment (rernst) [inline SMA.nf]: I would format like:
```
Sambamba_Filter_Condition(
    Samtools_Merge.out
    .map{ sample_id, bam_file -> bam_file }
    .combine(Sambamba_Index_Merge.out.map{ sample_id, bai_file -> bai_file })
)

```

File: SMA.nf [hunk 30]
```diff
@@ -0,0 +1,375 @@
...
+        GetPhaseSet_Bed(GATK_VariantFiltration_Bed.out)
+
+        // Split BAM into single haplotypes (BED approach)
+        Sambamba_Filter_Haplotype_Phaseset_Bed(Whatshap_Haplotag_Target_Bed.out.combine(
+            Sambamba_Index_Target_Bed.out.map{sample_id, bai_file -> bai_file}
+        ).combine(
+            ploidy_list
+        ).combine(
+           GetPhaseSet_Bed.out
+        ))
```
Comment (rernst) [inline SMA.nf]:
```
Sambamba_Filter_Haplotype_Phaseset_Bed(
    Whatshap_Haplotag_Target_Bed.out
    .combine(
        Sambamba_Index_Target_Bed.out.map{sample_id, bai_file -> bai_file}
    )
    .combine(ploidy_list)
    .combine(GetPhaseSet_Bed.out)
)
```

File: SMA.nf [hunk 31]
```diff
@@ -0,0 +1,375 @@
...
+        GetPhaseSet_Region(GATK_VariantFiltration_Region.out)
+
+        // Split BAM into single haplotypes (REGION approach)
+        Sambamba_Filter_Haplotype_Phaseset_Region(Whatshap_Haplotag_Target_Region.out.combine(
+            Sambamba_Index_Target_Region.out.map{sample_id, bai_file -> bai_file}
+        ).combine(
+            ploidy_list
+        ).combine(
+           GetPhaseSet_Region.out
+        ))
```
Comment (rernst) [inline SMA.nf]: See comment line 246

File: assets/get_phaseset.py [hunk 32]
```diff
@@ -0,0 +1,47 @@
+#! /usr/bin/env python3
+
+import argparse
+import vcfpy
+
+
+def ParseVCF(input_vcf, region):
+    chrom = region.split(":")[0]
+    start = int(region.split(":")[1].split("-")[0])
+    stop = int(region.split(":")[1].split("-")[1])
```
Comment (rernst) [inline assets/get_phaseset.py]:
```python
chrom, start, stop = re.split(':|-', region)

```

File: assets/get_phaseset.py [hunk 33]
```diff
@@ -0,0 +1,47 @@
...
+
+    ps = []
+    reader = vcfpy.Reader.from_path(input_vcf)
+    for record in reader.fetch(chrom, begin=start, end=stop):
```
Comment (rernst) [inline assets/get_phaseset.py]: vcfpy can use samtools style regions:
``` 
:param str chrom_or_region: name of the chromosome to jump to if
            begin and end are given and a samtools region string otherwise
            (e.g. "chr1:123,456-123,900").
```

File: assets/get_phaseset.py [hunk 34]
```diff
@@ -0,0 +1,47 @@
...
+import argparse
+import vcfpy
+
+
+def ParseVCF(input_vcf, region):
+    chrom = region.split(":")[0]
+    start = int(region.split(":")[1].split("-")[0])
+    stop = int(region.split(":")[1].split("-")[1])
+
+    ps = []
```
Comment (rernst) [inline assets/get_phaseset.py]: ps = `phasesets`?

File: assets/get_phaseset.py [hunk 35]
```diff
@@ -0,0 +1,47 @@
...
+    reader = vcfpy.Reader.from_path(input_vcf)
+    for record in reader.fetch(chrom, begin=start, end=stop):
+        phaseset = record.calls[0].data.get('PS')  # Assuming single sample VCF
+        if phaseset:
+            ps.append(phaseset)
+    if len(set(ps)) == 1:
```
Comment (rernst) [inline assets/get_phaseset.py]: Add newline to seperate block.

File: assets/get_phaseset.py [hunk 36]
```diff
@@ -0,0 +1,47 @@
...
+        return list(set(ps))[0]
+    else:  # if not unique ps is detected in roi, pick most prevelant
+        count_dic = {}
+        total_count = 0
+        for ps_group in ps:
+            if ps_group not in count_dic:
+                count_dic[ps_group] = 1
+            else:
+                count_dic[ps_group] += 1
```
Comment (rernst) [inline assets/get_phaseset.py]:
```python
if ps_group not in count_dic:
    count_dic[ps_group] = 0
count_dic[ps_group] += 1
```

File: assets/get_phaseset.py [hunk 37]
```diff
@@ -0,0 +1,47 @@
...
+            total_count += 1
+        for ps_group in count_dic:
+            if count_dic[ps_group]/total_count > args.freq:
+                return ps_group
```
Comment (rernst) [inline assets/get_phaseset.py]: This will not necessarily return the most prevalent? Just the first one which passes a certain frequency?

File: assets/get_phaseset.py [hunk 38]
```diff
@@ -0,0 +1,47 @@
+#! /usr/bin/env python3
+
+import argparse
+import vcfpy
+
+
+def ParseVCF(input_vcf, region):
```
Comment (rernst) [inline assets/get_phaseset.py]: Python functions are never in CamelCase. I would also rename it to something like: `get_phaseset_from_vcf`.

File: run_nextflow_sma_umcu.sh [hunk 39]
```diff
@@ -0,0 +1,91 @@
...
+    #  echo "Zip work directory"
+    find work -type f | egrep "\.(command|exitcode)" | zip -@ -q work.zip
+
+    #  echo "Remove work directory"
+    rm -r work
+
+    #  echo "Creating md5sum"
+    find -type f -not -iname 'md5sum.txt' -exec md5sum {} \; > md5sum.txt
+
+    echo "WES workflow completed successfully."
```
Comment (rernst) [inline run_nextflow_sma_umcu.sh]: WES!

File: run_nextflow_sma_umcu.sh [hunk 40]
```diff
@@ -0,0 +1,91 @@
+#!/bin/bash
```
Comment (rernst) [inline run_nextflow_sma_umcu.sh:1]: This script contains a lot of UMCU / slurm specific things. I am not sure whether I would add it like this or maybe even remove it... Just provide the nextflow commands, which should be way easier to reproduce in any environment.

**Review by rernst — COMMENTED** [review]
# Modules:
- Why detached/copied from van NextflowModules?
- If you copy code under license you typically need to include the license with the copied code: https://github.com/UMCUGenetics/NextflowModules/blob/master/LICENSE
  - `The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.`
- No need to add things like 'variantcaller' or 'phase' if the command does only 1 thing.

# Nextflow:
- Double newlines?

# Other:
- Add a license to the repo, without a license no one is allowed to use your code (copyright). 
- I think there is too much UMCU specific environment code in the repo. Examples: install script, run script, slurm configs. I am afraid that people that want to reuse this workflow outside the UMCU get stuck in our specific configurations. Would be better to provide a plain nextflow workflow which they should be able to run after reading in on the needed nextflow settings for their environment.

File: run_nextflow_sma_umcu.sh (hunk 39 above)
Comment (melferink) [reply run_nextflow_sma_umcu.sh]: WES rules!

File: Modules/Utils/GuppyBasecalling.nf (hunk 8 above)
Comment (melferink) [reply Modules/Utils/GuppyBasecalling.nf:1]: Left out of Modules as Guppy is not opensource/ containerized. In that sense, tool/version folder for guppy is different that other tools.

File: assets/get_phaseset.py (hunk 37 above)
Comment (melferink) [reply assets/get_phaseset.py]: Fair point. Luckily (for our analysis), this is never the case if freq > 0.50 (which is default 0.65).
Nevertheless, changed the code to make sure the most prevalent phaseset is tested.

File: assets/get_phaseset.py (hunk 33 above)
Comment (melferink) [reply assets/get_phaseset.py]: Well, that's convenient :)

File: README.md [hunk 41]
```diff
@@ -1,14 +1,25 @@
...
+* Rebase(calling) method using Guppy needs to be run on GPU compute node with sufficient resources for the non-gpu part of the workflow (cpu = 10, memory = 48GB). Resources can be lowered in SMA.config, but might result in longer analysis time.
+* Guppy needs to be installed locally.
+* Singularity(/Docker) needs to be installed on the system to run the workflow. Testing has been performed using Singularity 1.3.1.
+* OpenJDK and nextflow need to be installed (see install.sh script). Workflow has been tested with nextflow 23.10.0 and jdk21.0.2
+
+Note:
+* Workflow testing has been performed on UMC Utrecht HPC enviroment using SMA.config + SMA_umcu.config in wrapper run_nextflow_sma_umcu.sh.
+* Worfklow testing has only been performed using R9.4.1 sequencing data with guppy_6.1.2 basecalling.
+* General use of the workflow is described below using SMA.config, although some configuration might be needed dependent on your local infrastructure.
+* Reference genome need to be indexed. For more details see below in section "how to index the reference genome".
```
Comment (rernst) [inline README.md]: Reference genome **needs** to be indexed

File: README.md [hunk 42]
```diff
@@ -1,14 +1,25 @@
...
+* Minimal resource requirement for current SMA.config is cpu = 10 and memory = 48GB. These can be lowered in SMA.config, but might result in longer analysis time.
+* Rebase(calling) method using Guppy needs to be run on GPU compute node with sufficient resources for the non-gpu part of the workflow (cpu = 10, memory = 48GB). Resources can be lowered in SMA.config, but might result in longer analysis time.
+* Guppy needs to be installed locally.
+* Singularity(/Docker) needs to be installed on the system to run the workflow. Testing has been performed using Singularity 1.3.1.
+* OpenJDK and nextflow need to be installed (see install.sh script). Workflow has been tested with nextflow 23.10.0 and jdk21.0.2
+
+Note:
+* Workflow testing has been performed on UMC Utrecht HPC enviroment using SMA.config + SMA_umcu.config in wrapper run_nextflow_sma_umcu.sh.
+* Worfklow testing has only been performed using R9.4.1 sequencing data with guppy_6.1.2 basecalling.
+* General use of the workflow is described below using SMA.config, although some configuration might be needed dependent on your local infrastructure.
```
Comment (rernst) [inline README.md]: dependent -> depending

File: README.md [hunk 43]
```diff
@@ -17,45 +28,66 @@ SMA.config:
...
-<pre>
-  * rebase              start from FAST5 raw data ouput folder and include re-basecalling.
-  * bam                 start from Guppy folder including bam files.
-  * bam_remap           start from Guppy folder including bam files, but perform remapping to genome in config.
-  * bam_single          start from single bam file (should be only bam in the folder) without any Guppy/ONT information.
-  * bam_single_remap    start from single bam file (should be only bam in the folder) without any Guppy/ONT information and perform remapping.
-</pre>
+### --start and --input_path
+
+NextflowSMA can be started using different options (__rebase__, __bam__, __bam_single__) which require different input paths and assumes different files within these folders.
```
Comment (rernst) [inline README.md]: assumes -> assume?

File: README.md [hunk 44]
```diff
@@ -17,45 +28,66 @@ SMA.config:
...
+
+Optional for start __bam__ and __bam_single__ is the addition of ___remap__, in which reads will be remapped to the reference genome provided in SMA.config.\
```
Comment (rernst) [inline README.md]: Optionally the suffix ___remap__ can be added to  __bam__ and __bam_single__, this will enable remapping of the reads to the reference genome provided in SMA.config.

File: README.md [hunk 45]
```diff
@@ -17,45 +28,66 @@ SMA.config:
...
+To prevent downstream issues in variantcalling we would advise to always include ___remap__ unless it is certain that the reference genome in SMA.config has also been used for mappping reads in the input BAM.
```
Comment (rernst) [inline README.md]: variant calling

File: README.md [hunk 46]
```diff
@@ -17,45 +28,66 @@ SMA.config:
...
+    * start from BAM files within the Guppy output folder
+    * input_path = full path to Guppy output folder containing BAM files in the pass/ folder and sequencing_summary.txt in the root folder
+    
+  * __bam_remap__
+    * start from BAM files within the Guppy output folder and perform remapping
+    * input_path = full path to Guppy output folder containing BAM files in the pass/ folder and sequencing_summary.txt in the root folder
+     
+  * __bam_single__
+    * start from single bam file 
+    * input_path = full path to folder containing (only) one BAM file and corresponding BAI index file
```
Comment (rernst) [inline README.md]: Does the `bai` filename need to be in a specific format? for example: `.bam.bai`?

File: README.md [hunk 47]
```diff
@@ -17,45 +28,66 @@ SMA.config:
...
+    
+  * __bam_single_remap__
+    * start from single bam file and perform remapping
+    * input_path = full path to folder containing (only) one BAM file and corresponding BAI index file
+     
+    
+### --ploidy
+* Estimated copy number of SMN1 + SMN2 (integer)
+
+## how to index the reference genome.
```
Comment (rernst) [inline README.md]: **H**ow

Comment (rernst) [reply README.md]: And maybe add a small descriptive sentence?

The reference genome in `SMA.config` must be indexed using samtools, Picard and minimap2. Make sure that the index files are present in the same directory as the reference genome.

File: README.md [hunk 48]
```diff
@@ -17,45 +28,66 @@ SMA.config:
...
+```bash
+minimap2 -d <path_to_reference_fasta>.mmi <path_to_reference_fasta>
+```
 
-| Method | Description | Format |
-| :--- | :--- | :--- |
-|\<SMN2 copy number\>|SMN2 copynumber (ploidy)|\<int\>|
+# UMC Utrecht specific files
+SMA_umcu.config and run_nextflow_sma_umcu.sh are custom scripts to run the workflow on the UMC Utrecht HPC infrastructure.
+Please do no use these, although they could serve as a template for your specific system.
```
Comment (rernst) [inline README.md]: system -> compute environment.

File: README.md [hunk 49]
```diff
@@ -1,2 +1,93 @@
...
+* Singularity(/Docker) needs to be installed on the system to run the workflow. Testing has been performed using Singularity 1.3.1.
+* OpenJDK and nextflow need to be installed (see install.sh script). Workflow has been tested with nextflow 23.10.0 and jdk21.0.2
+
+Note:
+* Workflow testing has been performed on UMC Utrecht HPC enviroment using SMA.config + SMA_umcu.config in wrapper run_nextflow_sma_umcu.sh.
+* Worfklow testing has only been performed using R9.4.1 sequencing data with guppy_6.1.2 basecalling.
+* General use of the workflow is described below using SMA.config, although some configuration might be needed dependent on your local infrastructure.
+* Reference genome need to be indexed. For more details see below in section "how to index the reference genome".
+
+## Install OpenJDK and Nextflow within tools/ folder. Both are needed to run this workflow.
```
Comment (rernst) [inline README.md]:
```markdown
## Install  OpenJDK and Nextflow
The `install.sh` bash script can be used the install the OpenJDK and Nextflow versions used to develop this workflow. The scripts assumes to be run in the root folder of this repository and will create a tools folder to install both tools.

...

```

File: README.md [hunk 50]
```diff
@@ -1,2 +1,93 @@
...
+
+  Only needed in case of the rebase(calling) method:
+  genome_mapping_index          full path to reference genome minimap2 index (.mmi)
+  guppy_basecaller_path 	full path to guppy_basecaller executable
+  guppy_basecaller_config 	full path to basecalling model file to be used (.cfg)
+</pre>
+
+## Running SMA workflow
+```bash
+nextflow run SMA.nf -c SMA.config --input_path <input_path> --outdir <output_dir_path> --start <start> --ploidy <ploidy> --email <email>
```
Comment (rernst) [inline README.md]: I think you should/can refer to the tools installed using `install.sh`?

```bash
workflow_path='/change/to/repository/path'
export NXF_JAVA_HOME='$workflow_path/tools/java/jdk'
export NXF_VER=23.10.0

$workflow_path/tools/nextflow ...

```

**Review by rernst — COMMENTED** [review]
Comments on readme

File: README.md (hunk 46 above)
Comment (melferink) [reply README.md]: Actually.. indexing is always performed in workflow for bam_single, so removed from readme

File: SMA.nf (hunk 28 above)
Comment (melferink) [reply SMA.nf:123]: Will make ticket out of this.

**Review by rernst — APPROVED** [review]
